sql/idempotent/005-vectorizer.sql

```diff
@@ -0,0 +1,1178 @@
+
+    -- validate the formatting config
+    case formatting operator(pg_catalog.->>) 'implementation'
+        when 'python_template' then
+            formatting = ai._validate_formatting_python_template
+            ( formatting
+            , _source_schema
+            , _source_table
+            );
+        else
+            raise exception 'unrecognized formatting config';
+    end case;
+
+    -- validate the scheduling config
+    case scheduling operator(pg_catalog.->>) 'implementation'
+        when 'timescaledb' then
+            -- ok
+        when 'pg_cron' then
+            -- ok
+        when 'none' then
+            -- ok
+        else
+            raise exception 'unrecognized scheduling config';
+    end case;
+
+    -- grant select on source table to _grant_to roles
+    if grant_to is not null then
+        select pg_catalog.format
+        ( $sql$grant select on %I.%I to %s$sql$
+        , _source_schema
+        , _source_table
+        , (
+            select pg_catalog.string_agg(pg_catalog.quote_ident(x), ', ')
+            from pg_catalog.unnest(grant_to) x
+          )
+        ) into strict _sql;
+        execute _sql;
+    end if;
+
+    -- create the target table
+    perform ai._vectorizer_create_target_table
+    ( _source_schema
+    , _source_table
+    , _source_pk
+    , target_schema
+    , target_table
+    , _dimensions
+    );
+
+    -- grant select, insert, update on target table to _grant_to roles
+    if grant_to is not null then
+        select pg_catalog.format
+        ( $sql$grant select, insert, update on %I.%I to %s$sql$
+        , target_schema
+        , target_table
+        , (
+            select pg_catalog.string_agg(pg_catalog.quote_ident(x), ', ')
+            from pg_catalog.unnest(grant_to) x
+          )
+        ) into strict _sql;
+        execute _sql;
+    end if;
+
+    -- create queue table
+    perform ai._vectorizer_create_queue_table
+    ( queue_schema
+    , queue_table
+    , _source_pk
+    );
+
+    -- grant select, update, delete on queue table to _grant_to roles
+    if grant_to is not null then
+        select pg_catalog.format
+        ( $sql$grant select, update, delete on %I.%I to %s$sql$
+        , queue_schema
+        , queue_table
+        , (
+            select pg_catalog.string_agg(pg_catalog.quote_ident(x), ', ')
+            from pg_catalog.unnest(grant_to) x
+          )
+        ) into strict _sql;
+        execute _sql;
+    end if;
+
+    -- create trigger on source table to populate queue
+    perform ai._vectorizer_create_source_trigger
+    ( _trigger_name
+    , queue_schema
+    , queue_table
+    , _source_schema
+    , _source_table
+    , _source_pk
+    );
+
+    -- create view
+    perform ai._vectorizer_create_view
+    ( view_schema
+    , view_name
+    , _source_schema
+    , _source_table
+    , _source_pk
+    , target_schema
+    , target_table
+    );
+
+    -- grant select view to _grant_to roles
```

Comment:
why does the worker need this?

Reply:
the worker does not, but if the user adds a different kind of role to the _grant_to list, then they might need it